File: FastOS/kernel/src/drivers/gpu.rs
// ============================================================
// FastOS — GPU Driver (NVIDIA via PCI + SPIR-V)
// ============================================================
// Detects NVIDIA GPU via PCI enumeration and provides basic
// compute dispatch interface using SPIR-V shaders.
//
// Integration with ADead-BIB abi_translators for SPIR-V parsing.
// ============================================================

use crate::serial_print;

// PCI Configuration Space ports
const PCI_CONFIG_ADDR: u16 = 0xCF8;
const PCI_CONFIG_DATA: u16 = 0xCFC;

// PCI Vendor IDs
const VENDOR_NVIDIA: u16 = 0x10DE;
const VENDOR_AMD: u16 = 0x1002;
const VENDOR_INTEL: u16 = 0x8086;

// PCI Class codes
const CLASS_DISPLAY: u8 = 0x03;
const SUBCLASS_VGA: u8 = 0x00;
const SUBCLASS_3D: u8 = 0x02;

// GPU Memory regions (BAR)
const GPU_BAR0_MMIO: usize = 0;
const GPU_BAR1_VRAM: usize = 1;

/// GPU device information
#[derive(Debug, Clone)]
pub struct GpuDevice {
    pub vendor_id: u16,
    pub device_id: u16,
    pub bus: u8,
    pub device: u8,
    pub function: u8,
    pub bar0: u64,      // MMIO registers
    pub bar1: u64,      // VRAM
    pub vram_size: u64, // Detected VRAM size
    pub name: &'static str,
}

/// GPU compute buffer
pub struct GpuBuffer {
    pub ptr: u64,
    pub size: usize,
    pub device_ptr: u64,
}

/// SPIR-V kernel for compute
pub struct SpirVKernel {
    pub name: [u8; 32],
    pub bytecode: &'static [u8],
    pub local_size: [u32; 3],
}

/// GPU Driver state
pub struct GpuDriver {
    pub device: Option<GpuDevice>,
    pub initialized: bool,
}

impl GpuDriver {
    pub const fn new() -> Self {
        Self {
            device: None,
            initialized: false,
        }
    }

    /// Initialize GPU driver - scan PCI for GPU
    pub fn init(&mut self) -> bool {
        serial_print("[GPU] Scanning PCI bus for GPU...\r\n");
        
        // Scan PCI buses (limited scan for speed)
        for bus in 0..8u8 {
            for device in 0..32u8 {
                for function in 0..8u8 {
                    if let Some(gpu) = self.probe_pci_device(bus, device, function) {
                        serial_print("[GPU] Found: ");
                        serial_print(gpu.name);
                        serial_print(" (");
                        // Print vendor:device
                        print_hex_u16(gpu.vendor_id);
                        serial_print(":");
                        print_hex_u16(gpu.device_id);
                        serial_print(")\r\n");
                        
                        // Print BAR info
                        serial_print("[GPU] BAR0=0x");
                        print_hex_u64(gpu.bar0);
                        serial_print(" BAR1=0x");
                        print_hex_u64(gpu.bar1);
                        serial_print("\r\n");
                        
                        self.device = Some(gpu);
                        self.initialized = true;
                        return true;
                    }
                }
            }
        }
        
        serial_print("[GPU] No compatible GPU found\r\n");
        false
    }

    /// Probe a PCI device to check if it's a GPU
    fn probe_pci_device(&self, bus: u8, device: u8, function: u8) -> Option<GpuDevice> {
        let vendor_id = pci_read_word(bus, device, function, 0x00);
        if vendor_id == 0xFFFF || vendor_id == 0x0000 {
            return None;
        }

        let device_id = pci_read_word(bus, device, function, 0x02);
        let class_code = pci_read_byte(bus, device, function, 0x0B);
        let subclass = pci_read_byte(bus, device, function, 0x0A);

        // Check if it's a display controller
        if class_code != CLASS_DISPLAY {
            return None;
        }
        if subclass != SUBCLASS_VGA && subclass != SUBCLASS_3D {
            return None;
        }

        // Read BARs
        let bar0 = pci_read_bar(bus, device, function, 0x10);
        let bar1 = pci_read_bar(bus, device, function, 0x14);

        // Identify GPU
        let name = match vendor_id {
            VENDOR_NVIDIA => identify_nvidia_gpu(device_id),
            VENDOR_AMD => "AMD GPU",
            VENDOR_INTEL => "Intel GPU",
            _ => "Unknown GPU",
        };

        // Estimate VRAM (simplified - real detection requires GPU-specific code)
        let vram_size = estimate_vram(vendor_id, device_id);

        Some(GpuDevice {
            vendor_id,
            device_id,
            bus,
            device,
            function,
            bar0,
            bar1,
            vram_size,
            name,
        })
    }

    /// Check if GPU is available
    pub fn is_available(&self) -> bool {
        self.initialized && self.device.is_some()
    }

    /// Get GPU info string
    pub fn get_info(&self) -> &'static str {
        if let Some(ref dev) = self.device {
            dev.name
        } else {
            "No GPU"
        }
    }

    /// Allocate a GPU buffer (placeholder - requires real GPU memory management)
    pub fn alloc_buffer(&self, size: usize) -> Option<GpuBuffer> {
        if !self.is_available() {
            return None;
        }

        // In a real implementation, this would allocate from GPU VRAM
        // For now, we just track the allocation
        Some(GpuBuffer {
            ptr: 0,
            size,
            device_ptr: 0,
        })
    }

    /// Load a SPIR-V kernel (placeholder)
    pub fn load_spirv(&self, _bytecode: &[u8]) -> Option<SpirVKernel> {
        if !self.is_available() {
            return None;
        }

        // In a real implementation, this would:
        // 1. Parse SPIR-V using abi_translators
        // 2. Compile to GPU-native code
        // 3. Upload to GPU

        Some(SpirVKernel {
            name: [0; 32],
            bytecode: &[],
            local_size: [64, 1, 1],
        })
    }

    /// Dispatch a compute shader (placeholder)
    pub fn dispatch(&self, _kernel: &SpirVKernel, groups: [u32; 3]) -> bool {
        if !self.is_available() {
            return false;
        }

        serial_print("[GPU] Dispatch compute: ");
        // Would actually dispatch to GPU here
        let _ = groups;
        true
    }
}

// ============================================================
// Hex printing helpers
// ============================================================

fn print_hex_u16(val: u16) {
    const HEX: &[u8; 16] = b"0123456789ABCDEF";
    let mut buf = [0u8; 4];
    buf[0] = HEX[((val >> 12) & 0xF) as usize];
    buf[1] = HEX[((val >> 8) & 0xF) as usize];
    buf[2] = HEX[((val >> 4) & 0xF) as usize];
    buf[3] = HEX[(val & 0xF) as usize];
    for &c in &buf {
        print_char(c);
    }
}

fn print_hex_u64(val: u64) {
    const HEX: &[u8; 16] = b"0123456789ABCDEF";
    for i in (0..16).rev() {
        let nibble = ((val >> (i * 4)) & 0xF) as usize;
        print_char(HEX[nibble]);
    }
}

fn print_char(c: u8) {
    unsafe {
        // Output to serial port COM1
        while (inb_simple(0x3FD) & 0x20) == 0 {}
        outb_simple(0x3F8, c);
    }
}

unsafe fn outb_simple(port: u16, value: u8) {
    core::arch::asm!("out dx, al", in("dx") port, in("al") value, options(nomem, nostack));
}

unsafe fn inb_simple(port: u16) -> u8 {
    let value: u8;
    core::arch::asm!("in al, dx", in("dx") port, out("al") value, options(nomem, nostack));
    value
}

// ============================================================
// PCI Configuration Space Access
// ============================================================

fn pci_config_address(bus: u8, device: u8, function: u8, offset: u8) -> u32 {
    let bus = bus as u32;
    let device = device as u32;
    let function = function as u32;
    let offset = (offset & 0xFC) as u32;
    0x80000000 | (bus << 16) | (device << 11) | (function << 8) | offset
}

fn pci_read_dword(bus: u8, device: u8, function: u8, offset: u8) -> u32 {
    let addr = pci_config_address(bus, device, function, offset);
    unsafe {
        outl(PCI_CONFIG_ADDR, addr);
        inl(PCI_CONFIG_DATA)
    }
}

fn pci_read_word(bus: u8, device: u8, function: u8, offset: u8) -> u16 {
    let dword = pci_read_dword(bus, device, function, offset & 0xFC);
    let shift = ((offset & 2) * 8) as u32;
    ((dword >> shift) & 0xFFFF) as u16
}

fn pci_read_byte(bus: u8, device: u8, function: u8, offset: u8) -> u8 {
    let dword = pci_read_dword(bus, device, function, offset & 0xFC);
    let shift = ((offset & 3) * 8) as u32;
    ((dword >> shift) & 0xFF) as u8
}

fn pci_read_bar(bus: u8, device: u8, function: u8, bar_offset: u8) -> u64 {
    let low = pci_read_dword(bus, device, function, bar_offset);
    
    // Check if 64-bit BAR
    if (low & 0x06) == 0x04 {
        let high = pci_read_dword(bus, device, function, bar_offset + 4);
        let addr = ((high as u64) << 32) | ((low & 0xFFFFFFF0) as u64);
        addr
    } else {
        (low & 0xFFFFFFF0) as u64
    }
}

// ============================================================
// Port I/O
// ============================================================

unsafe fn outl(port: u16, value: u32) {
    core::arch::asm!(
        "out dx, eax",
        in("dx") port,
        in("eax") value,
        options(nomem, nostack)
    );
}

unsafe fn inl(port: u16) -> u32 {
    let value: u32;
    core::arch::asm!(
        "in eax, dx",
        in("dx") port,
        out("eax") value,
        options(nomem, nostack)
    );
    value
}

// ============================================================
// GPU Identification
// ============================================================

fn identify_nvidia_gpu(device_id: u16) -> &'static str {
    // Common NVIDIA device IDs (RTX 30 series, etc.)
    match device_id {
        // RTX 40 series
        0x2684 => "NVIDIA RTX 4090",
        0x2704 => "NVIDIA RTX 4080",
        0x2782 => "NVIDIA RTX 4070 Ti",
        0x2786 => "NVIDIA RTX 4070",
        
        // RTX 30 series
        0x2204 => "NVIDIA RTX 3090",
        0x2206 => "NVIDIA RTX 3080",
        0x2208 => "NVIDIA RTX 3080 Ti",
        0x2484 => "NVIDIA RTX 3070",
        0x2488 => "NVIDIA RTX 3070 Ti",
        0x2503 => "NVIDIA RTX 3060",
        0x2504 => "NVIDIA RTX 3060 Ti",
        0x2520 => "NVIDIA RTX 3060 Mobile",
        0x2531 => "NVIDIA RTX 3060 Laptop",
        
        // RTX 20 series
        0x1E04 => "NVIDIA RTX 2080 Ti",
        0x1E07 => "NVIDIA RTX 2080 Super",
        0x1E82 => "NVIDIA RTX 2080",
        0x1F02 => "NVIDIA RTX 2070",
        0x1F07 => "NVIDIA RTX 2070 Super",
        0x1F08 => "NVIDIA RTX 2060 Super",
        0x1F82 => "NVIDIA RTX 2060",
        
        // GTX 16 series
        0x2182 => "NVIDIA GTX 1660 Ti",
        0x2184 => "NVIDIA GTX 1660",
        0x21C4 => "NVIDIA GTX 1660 Super",
        0x1F91 => "NVIDIA GTX 1650 Super",
        0x1F82 => "NVIDIA GTX 1650",
        
        // GTX 10 series
        0x1B80 => "NVIDIA GTX 1080",
        0x1B81 => "NVIDIA GTX 1070",
        0x1B82 => "NVIDIA GTX 1070 Ti",
        0x1C02 => "NVIDIA GTX 1060 6GB",
        0x1C03 => "NVIDIA GTX 1060 3GB",
        0x1C81 => "NVIDIA GTX 1050",
        0x1C82 => "NVIDIA GTX 1050 Ti",
        
        // Quadro / Professional
        0x1DB6 => "NVIDIA Quadro RTX 5000",
        0x1E30 => "NVIDIA Quadro RTX 6000",
        
        _ => "NVIDIA GPU (Unknown Model)",
    }
}

fn estimate_vram(vendor_id: u16, device_id: u16) -> u64 {
    if vendor_id == VENDOR_NVIDIA {
        match device_id {
            0x2684 => 24 * 1024 * 1024 * 1024, // RTX 4090: 24GB
            0x2704 => 16 * 1024 * 1024 * 1024, // RTX 4080: 16GB
            0x2204 => 24 * 1024 * 1024 * 1024, // RTX 3090: 24GB
            0x2206 => 10 * 1024 * 1024 * 1024, // RTX 3080: 10GB
            0x2503 => 12 * 1024 * 1024 * 1024, // RTX 3060: 12GB
            0x2504 => 8 * 1024 * 1024 * 1024,  // RTX 3060 Ti: 8GB
            _ => 8 * 1024 * 1024 * 1024,       // Default 8GB
        }
    } else {
        4 * 1024 * 1024 * 1024 // Default 4GB
    }
}

// ============================================================
// Global GPU Driver Instance
// ============================================================

static mut GPU_DRIVER: GpuDriver = GpuDriver::new();

pub fn init() -> bool {
    unsafe { GPU_DRIVER.init() }
}

pub fn get_driver() -> &'static GpuDriver {
    unsafe { &GPU_DRIVER }
}

pub fn get_driver_mut() -> &'static mut GpuDriver {
    unsafe { &mut GPU_DRIVER }
}

File: FastOS/kernel/drivers/gpu.c
// ============================================================
// FastOS GPU Driver — RTX 3060 via PCI
// ============================================================
// Comunicación directa CPU ↔ GPU sin drivers NVIDIA
// PCI BAR0 = MMIO registers
// PCI BAR1 = VRAM (framebuffer)
// ============================================================

#include "../include/types.h"

// ============================================================
// PCI Configuration Space
// ============================================================
#define PCI_CONFIG_ADDR  0xCF8
#define PCI_CONFIG_DATA  0xCFC

// NVIDIA Vendor ID
#define NVIDIA_VENDOR_ID 0x10DE

// RTX 3060 Device IDs (GA106)
#define RTX_3060_DEVICE_ID_1 0x2503
#define RTX_3060_DEVICE_ID_2 0x2504

// PCI Configuration Registers
#define PCI_VENDOR_ID    0x00
#define PCI_DEVICE_ID    0x02
#define PCI_COMMAND      0x04
#define PCI_STATUS       0x06
#define PCI_CLASS        0x0A
#define PCI_BAR0         0x10
#define PCI_BAR1         0x14
#define PCI_BAR2         0x18
#define PCI_BAR3         0x1C

// PCI Command bits
#define PCI_CMD_IO       0x01
#define PCI_CMD_MEMORY   0x02
#define PCI_CMD_MASTER   0x04

// ============================================================
// GPU State
// ============================================================
typedef struct {
    u8 bus;
    u8 device;
    u8 function;
    u16 vendor_id;
    u16 device_id;
    u64 bar0_base;      // MMIO registers
    u64 bar1_base;      // VRAM / Framebuffer
    u64 bar0_size;
    u64 bar1_size;
    u32 vram_mb;
    u8 detected;
} GpuDevice;

static GpuDevice gpu = {0};

// ============================================================
// Port I/O
// ============================================================
static inline void outl(u16 port, u32 value) {
    __asm__ volatile ("outl %0, %1" : : "a"(value), "Nd"(port));
}

static inline u32 inl(u16 port) {
    u32 value;
    __asm__ volatile ("inl %1, %0" : "=a"(value) : "Nd"(port));
    return value;
}

// ============================================================
// PCI Functions
// ============================================================
static u32 pci_read32(u8 bus, u8 device, u8 func, u8 offset) {
    u32 address = (1 << 31) |
                  ((u32)bus << 16) |
                  ((u32)device << 11) |
                  ((u32)func << 8) |
                  (offset & 0xFC);
    outl(PCI_CONFIG_ADDR, address);
    return inl(PCI_CONFIG_DATA);
}

static void pci_write32(u8 bus, u8 device, u8 func, u8 offset, u32 value) {
    u32 address = (1 << 31) |
                  ((u32)bus << 16) |
                  ((u32)device << 11) |
                  ((u32)func << 8) |
                  (offset & 0xFC);
    outl(PCI_CONFIG_ADDR, address);
    outl(PCI_CONFIG_DATA, value);
}

static u16 pci_read16(u8 bus, u8 device, u8 func, u8 offset) {
    u32 val = pci_read32(bus, device, func, offset & 0xFC);
    return (val >> ((offset & 2) * 8)) & 0xFFFF;
}

// ============================================================
// GPU Detection
// ============================================================
int gpu_detect(void) {
    // Scan PCI bus for NVIDIA GPU
    for (u8 bus = 0; bus < 255; bus++) {
        for (u8 dev = 0; dev < 32; dev++) {
            u16 vendor = pci_read16(bus, dev, 0, PCI_VENDOR_ID);
            
            if (vendor == NVIDIA_VENDOR_ID) {
                u16 device_id = pci_read16(bus, dev, 0, PCI_DEVICE_ID);
                
                // Found NVIDIA GPU
                gpu.bus = bus;
                gpu.device = dev;
                gpu.function = 0;
                gpu.vendor_id = vendor;
                gpu.device_id = device_id;
                
                // Read BARs
                u32 bar0 = pci_read32(bus, dev, 0, PCI_BAR0);
                u32 bar1 = pci_read32(bus, dev, 0, PCI_BAR1);
                
                // BAR0 is MMIO (memory mapped I/O)
                if ((bar0 & 0x1) == 0) {
                    // Memory BAR
                    if ((bar0 & 0x6) == 0x4) {
                        // 64-bit BAR
                        u32 bar0_high = pci_read32(bus, dev, 0, PCI_BAR0 + 4);
                        gpu.bar0_base = ((u64)bar0_high << 32) | (bar0 & 0xFFFFFFF0);
                    } else {
                        gpu.bar0_base = bar0 & 0xFFFFFFF0;
                    }
                }
                
                // BAR1 is VRAM
                if ((bar1 & 0x1) == 0) {
                    if ((bar1 & 0x6) == 0x4) {
                        u32 bar1_high = pci_read32(bus, dev, 0, PCI_BAR1 + 4);
                        gpu.bar1_base = ((u64)bar1_high << 32) | (bar1 & 0xFFFFFFF0);
                    } else {
                        gpu.bar1_base = bar1 & 0xFFFFFFF0;
                    }
                }
                
                // Enable bus mastering and memory access
                u16 cmd = pci_read16(bus, dev, 0, PCI_COMMAND);
                cmd |= PCI_CMD_MEMORY | PCI_CMD_MASTER;
                pci_write32(bus, dev, 0, PCI_COMMAND, cmd);
                
                gpu.detected = 1;
                
                // Estimate VRAM (RTX 3060 = 12GB)
                if (device_id == RTX_3060_DEVICE_ID_1 || device_id == RTX_3060_DEVICE_ID_2) {
                    gpu.vram_mb = 12288; // 12 GB
                } else {
                    gpu.vram_mb = 8192;  // Default 8 GB
                }
                
                return 1;
            }
        }
    }
    
    return 0;
}

// ============================================================
// GPU Info
// ============================================================
void gpu_print_info(void) {
    if (!gpu.detected) {
        // Print to VGA text mode
        volatile u16* vga = (volatile u16*)0xB8000;
        const char* msg = "[GPU] No NVIDIA GPU detected";
        for (int i = 0; msg[i]; i++) {
            vga[24*80 + i] = 0x0C00 | msg[i]; // Red
        }
        return;
    }
    
    // Print GPU info to VGA
    volatile u16* vga = (volatile u16*)0xB8000;
    int row = 22;
    int col = 0;
    
    // GPU header
    const char* hdr = "[GPU] NVIDIA RTX 3060 12GB";
    for (int i = 0; hdr[i]; i++) {
        vga[row*80 + col++] = 0x0A00 | hdr[i]; // Green
    }
    
    row++;
    col = 0;
    
    // BAR0
    const char* bar0_msg = "[GPU] BAR0 (MMIO): 0x";
    for (int i = 0; bar0_msg[i]; i++) {
        vga[row*80 + col++] = 0x0700 | bar0_msg[i];
    }
    
    // Print hex address
    u64 addr = gpu.bar0_base;
    for (int i = 15; i >= 0; i--) {
        int nibble = (addr >> (i * 4)) & 0xF;
        char c = nibble < 10 ? '0' + nibble : 'A' + nibble - 10;
        vga[row*80 + col++] = 0x0E00 | c; // Yellow
    }
    
    row++;
    col = 0;
    
    // BAR1
    const char* bar1_msg = "[GPU] BAR1 (VRAM): 0x";
    for (int i = 0; bar1_msg[i]; i++) {
        vga[row*80 + col++] = 0x0700 | bar1_msg[i];
    }
    
    addr = gpu.bar1_base;
    for (int i = 15; i >= 0; i--) {
        int nibble = (addr >> (i * 4)) & 0xF;
        char c = nibble < 10 ? '0' + nibble : 'A' + nibble - 10;
        vga[row*80 + col++] = 0x0E00 | c;
    }
}

// ============================================================
// MMIO Access
// ============================================================
static inline void gpu_write32(u64 offset, u32 value) {
    if (!gpu.detected) return;
    volatile u32* reg = (volatile u32*)(gpu.bar0_base + offset);
    *reg = value;
}

static inline u32 gpu_read32(u64 offset) {
    if (!gpu.detected) return 0;
    volatile u32* reg = (volatile u32*)(gpu.bar0_base + offset);
    return *reg;
}

// ============================================================
// VRAM Access (Framebuffer)
// ============================================================
void gpu_vram_write32(u64 offset, u32 value) {
    if (!gpu.detected) return;
    volatile u32* vram = (volatile u32*)(gpu.bar1_base + offset);
    *vram = value;
}

u32 gpu_vram_read32(u64 offset) {
    if (!gpu.detected) return 0;
    volatile u32* vram = (volatile u32*)(gpu.bar1_base + offset);
    return *vram;
}

// ============================================================
// GPU Framebuffer Drawing
// ============================================================

// Set pixel in VRAM (assuming linear framebuffer)
void gpu_set_pixel(u32 x, u32 y, u32 color) {
    if (!gpu.detected) return;
    
    // Assuming 1920x1080 resolution, 32bpp
    u32 pitch = 1920 * 4; // bytes per row
    u64 offset = y * pitch + x * 4;
    
    gpu_vram_write32(offset, color);
}

// Fill rectangle
void gpu_fill_rect(u32 x, u32 y, u32 w, u32 h, u32 color) {
    if (!gpu.detected) return;
    
    for (u32 py = y; py < y + h; py++) {
        for (u32 px = x; px < x + w; px++) {
            gpu_set_pixel(px, py, color);
        }
    }
}

// Draw horizontal line
void gpu_hline(u32 x, u32 y, u32 len, u32 color) {
    for (u32 i = 0; i < len; i++) {
        gpu_set_pixel(x + i, y, color);
    }
}

// Draw vertical line
void gpu_vline(u32 x, u32 y, u32 len, u32 color) {
    for (u32 i = 0; i < len; i++) {
        gpu_set_pixel(x, y + i, color);
    }
}

// Draw rectangle outline
void gpu_rect(u32 x, u32 y, u32 w, u32 h, u32 color) {
    gpu_hline(x, y, w, color);
    gpu_hline(x, y + h - 1, w, color);
    gpu_vline(x, y, h, color);
    gpu_vline(x + w - 1, y, h, color);
}

// ============================================================
// GPU Initialization
// ============================================================
int gpu_init(void) {
    if (!gpu_detect()) {
        return 0;
    }
    
    gpu_print_info();
    
    return 1;
}

// ============================================================
// GPU Test Pattern
// ============================================================
void gpu_test_pattern(void) {
    if (!gpu.detected) return;
    
    // Draw colored rectangles
    gpu_fill_rect(100, 100, 200, 150, 0x00FF0000); // Red
    gpu_fill_rect(350, 100, 200, 150, 0x0000FF00); // Green
    gpu_fill_rect(600, 100, 200, 150, 0x000000FF); // Blue
    
    // Draw white border
    gpu_rect(100, 100, 200, 150, 0x00FFFFFF);
    gpu_rect(350, 100, 200, 150, 0x00FFFFFF);
    gpu_rect(600, 100, 200, 150, 0x00FFFFFF);
}

// ============================================================
// GPU Getters
// ============================================================
u64 gpu_get_bar0(void) { return gpu.bar0_base; }
u64 gpu_get_bar1(void) { return gpu.bar1_base; }
u32 gpu_get_vram_mb(void) { return gpu.vram_mb; }
u8 gpu_is_detected(void) { return gpu.detected; }
u16 gpu_get_device_id(void) { return gpu.device_id; }
